### PR TITLE
Removed `final` keywords to add more flexibility.

### LINKED ISSUE #4: Final methods in BundleWithPlugins
Hi

Any reason `__construct()` / `build()` / `boot()` are marked as `final` in `BundleWithPlugins`?
This adds a huge limitation for regular bundles that would use this system as it cannot override those methods any more, e.g. to add compiler passes or anything that is done at boot time, not talking about the constructor.

Big projects like eZ could not use the plugin system because of this limitation.
I'd do a PR of course, but I wanted to discuss this first :-)

### PR DESCRIPTION
**Pull request by lolautruche**
`ExtensionWithPlugins` and `ConfigurationWithPlugins` can now be extended when a bundle has specific/complex dependencies and behavior. In that regard, I also changed some `private` properties to `protected` so that sub-classes can access them.
In the same context, I added shorthands methods to load plugins (`ExtensionWithPlugins`) and add plugin config (`ConfigurationWithPlugins`).

Fixes #4

### CODE REVIEW

**Comment by matthiasnoback** [comment]
@lolautruche Thanks! Though I'd like to keep the properties private. You shouldn't modify them in subclasses anyway. Could you revert that part and rebase, then I'll merge it right-away :)

**Comment by lolautruche** [comment]
@matthiasnoback Right. I did this before adding the shorthand methods, but now they're here we can keep them private indeed :smile:

**Comment by lolautruche** [comment]
However, I'd keep `$rootNodeName` protected in `ConfigurationWithPlugins`, as it still needs to be accessible for this class to be extendable. What do you think?

**Comment by lolautruche** [comment]
Updated, but keeping `$rootNodeName` protected.

**Comment by matthiasnoback** [comment]
To be honest, I don't want to make anything `protected` really. I've never liked `protected` things in open source software, because you're basically opening up your class internals and people will start depending on implementation details that you might want to change later on. The only thing people should depend upon is the public API, i.e. interfaces (the public parts of everything). The motto is: if you want to change something, use object composition, not inheritance. 

However, this library explicitly asks you to _extend_ from `BundleWithPlugins*, so the composition over inheritance rule doesn't apply anymore. Still, I'd like to keep everything from the classes in this library "private", so that I can change any implementation detail at any time.

In conclusion: I'd like this PR to just contain the removal of the `final` keywords :) If we have to open up the internals of the classes, then I'd like to discuss this per use case. Maybe there are alternatives which won't turn your own code into a mess ;) In the end, my goal is that this library is useful for others.

**Comment by lolautruche** [comment]
Fixed.

**Comment by matthiasnoback** [comment]
It seems like the fix wasn't pushed yet :)

**Comment by lolautruche** [comment]
Yes it was. Properties are private again :wink: 
Shorthand methods must still be protected, otherwise it's unusable when extending, specifically because `$registeredPlugins` is private. I could mark them as final though.

**Comment by matthiasnoback** [comment]
Oops, sorry man. However, when you have to extend, I'd say you just make calls to the parent functions, right? So if you extend `ExtensionWithPlugins` then in your `load()` method you just call `parent::load()`.

I've been thinking about this, and I don't think you should need to extend at all. After all, the `BundleWithPlugins` is just a convenient parent class, the real deal is in the `ExtensionWithPlugins`, but you can easily decorate it, like this:

``` php
class CustomExtension
{
    public function __construct(ExtensionInterface $extensionWithPlugins)
    {
        $this->extensionWithPlugins = $extensionWithPlugins;
    }

    public function load(...)
    {
        // custom load stuff
        $this->extensionWithPlugins->load(...);
    }
}
```

**Comment by lolautruche** [comment]
> So if you extend ExtensionWithPlugins then in your load() method you just call parent::load()

Yes, but then you can't have minimum control on the configuration, since it's processed in `ExtensionWithPlugins::load()`. So you won't be able to have _default_ configuration, or you'd have 2 configuration classes.

> I've been thinking about this, and I don't think you should need to extend at all. After all, the BundleWithPlugins is just a convenient parent class, the real deal is in the ExtensionWithPlugins, but you can easily decorate it

While I agree with decoration argument, my statement above remains as `ExtensionWithPlugins` and `ConfigurationWithPlugins` are tightly coupled.

**Comment by matthiasnoback** [comment]
Ah, I see. Sorry, I didn't grasp that part yet.

I'm also sorry that I can't be more welcoming to this change request. It seems like you would need it, but I also think that it sacrifices too much of the robustness of this library if we open up the API in this way.

If you need a default configuration, then you will end up duplicating half of the code in `ConfigurationWithPlugins` anyway. The same goes for `ExtensionWithPlugins`. You shouldn't need to call `loadPlugins()` yourself. If you use this library, then the true force of it lies in that you don't have to do anything yourself anymore.

**Comment by lolautruche** [comment]
@matthiasnoback Then this PR seems to be useless... As far as I can see, if we cannot at least call the plugins load/registration, we can then leave the `final` keyword.

Closing.

**Comment by matthiasnoback** [comment]
Thanks for your effort anyway! Again, apologies for the fact that we couldn't agree on a solution.
